File: ts/src/reviewer/chain.ts
import type { Reviewer } from "./protocol.ts";
import { SEVERITY_INFO, type CheckResult, type Verdict } from "./types.ts";

export type ChainStrategy = "strictest" | "majority" | "weighted";

export class ReviewerChain implements Reviewer {
  private readonly reviewers: Reviewer[];
  private readonly strategy: ChainStrategy;

  constructor(reviewers: Reviewer[], strategy: ChainStrategy = "strictest") {
    this.reviewers = reviewers;
    this.strategy = strategy;
  }

  async review(
    output: string,
    expected: string[],
    query: string,
  ): Promise<Verdict> {
    const verdicts: Verdict[] = [];
    for (const reviewer of this.reviewers) {
      verdicts.push(await reviewer.review(output, expected, query));
    }
    return this.resolve(verdicts);
  }

  private resolve(verdicts: Verdict[]): Verdict {
    if (verdicts.length === 0) {
      return {
        passed: true,
        reasoning: "No reviewers in chain",
        severity: SEVERITY_INFO,
        checks: [],
        score: 1.0,
      };
    }

    const allChecks: CheckResult[] = verdicts.flatMap((v) => v.checks);
    const avgScore =
      verdicts.reduce((sum, v) => sum + v.score, 0) / verdicts.length;

    switch (this.strategy) {
      case "strictest":
        return this.resolveStrictest(verdicts, allChecks, avgScore);
      case "majority":
        return this.resolveMajority(verdicts, allChecks, avgScore);
      case "weighted":
        return this.resolveWeighted(verdicts, allChecks);
      default:
        throw new Error(`Unknown strategy: ${this.strategy}`);
    }
  }

  private resolveStrictest(
    verdicts: Verdict[],
    allChecks: CheckResult[],
    avgScore: number,
  ): Verdict {
    const failed = verdicts.filter((v) => !v.passed);
    if (failed.length === 0) {
      return {
        passed: true,
        reasoning: "All reviewers passed",
        severity: SEVERITY_INFO,
        checks: allChecks,
        score: avgScore,
      };
    }
    const worst = failed.reduce((a, b) =>
      a.severity.weight >= b.severity.weight ? a : b,
    );
    const reasons = failed.map((v) => v.reasoning);
    return {
      passed: false,
      reasoning: `${failed.length} reviewer(s) failed: ${reasons.join("; ")}`,
      severity: worst.severity,
      checks: allChecks,
      score: avgScore,
    };
  }

  private resolveMajority(
    verdicts: Verdict[],
    allChecks: CheckResult[],
    avgScore: number,
  ): Verdict {
    const passCount = verdicts.filter((v) => v.passed).length;
    const passed = passCount > verdicts.length / 2;
    if (passed) {
      return {
        passed: true,
        reasoning: `Majority passed (${passCount}/${verdicts.length})`,
        severity: SEVERITY_INFO,
        checks: allChecks,
        score: avgScore,
      };
    }
    const failed = verdicts.filter((v) => !v.passed);
    const worst = failed.reduce((a, b) =>
      a.severity.weight >= b.severity.weight ? a : b,
    );
    return {
      passed: false,
      reasoning: `Majority failed (${failed.length}/${verdicts.length})`,
      severity: worst.severity,
      checks: allChecks,
      score: avgScore,
    };
  }

  private resolveWeighted(
    verdicts: Verdict[],
    allChecks: CheckResult[],
  ): Verdict {
    const totalWeight = verdicts.reduce(
      (sum, v) => sum + v.severity.weight,
      0,
    );
    if (totalWeight === 0) {
      return {
        passed: true,
        reasoning: "No weighted signal",
        severity: SEVERITY_INFO,
        checks: allChecks,
        score: 1.0,
      };
    }
    const weightedScore =
      verdicts.reduce((sum, v) => sum + v.score * v.severity.weight, 0) /
      totalWeight;
    const passed = weightedScore >= 0.5;
    const worstFailing = verdicts.filter((v) => !v.passed);
    const severity =
      worstFailing.length > 0
        ? worstFailing.reduce((a, b) =>
            a.severity.weight >= b.severity.weight ? a : b,
          ).severity
        : SEVERITY_INFO;
    return {
      passed,
      reasoning: `Weighted score: ${weightedScore.toFixed(2)}`,
      severity,
      checks: allChecks,
      score: weightedScore,
    };
  }
}

File: src/synapt_eval/reviewer/chain.py
"""ReviewerChain: composes multiple reviewers with conflict resolution."""

from __future__ import annotations

from typing import Any

from synapt_eval.reviewer.protocol import Reviewer
from synapt_eval.reviewer.types import (
    SEVERITY_INFO,
    Verdict,
)


class ReviewerChain(Reviewer):
    """Composes multiple reviewers with configurable conflict resolution.

    Strategies:
    - "strictest": any reviewer failure fails the chain (AND logic)
    - "majority": pass if >50% of reviewers pass
    - "weighted": weighted average of scores using severity weights
    """

    def __init__(
        self,
        reviewers: list[Reviewer],
        strategy: str = "strictest",
    ) -> None:
        self._reviewers = reviewers
        self._strategy = strategy

    async def review(
        self,
        output: str,
        expected: list[str],
        query: str,
        **kwargs: Any,
    ) -> Verdict:
        verdicts: list[Verdict] = []
        for reviewer in self._reviewers:
            verdict = await reviewer.review(output, expected, query, **kwargs)
            verdicts.append(verdict)
        return self._resolve(verdicts)

    def _resolve(self, verdicts: list[Verdict]) -> Verdict:
        if not verdicts:
            return Verdict(
                passed=True,
                reasoning="No reviewers in chain",
                severity=SEVERITY_INFO,
            )

        all_checks = [c for v in verdicts for c in v.checks]
        avg_score = sum(v.score for v in verdicts) / len(verdicts)

        if self._strategy == "strictest":
            return self._resolve_strictest(verdicts, all_checks, avg_score)
        elif self._strategy == "majority":
            return self._resolve_majority(verdicts, all_checks, avg_score)
        elif self._strategy == "weighted":
            return self._resolve_weighted(verdicts, all_checks)
        else:
            raise ValueError(f"Unknown strategy: {self._strategy}")

    def _resolve_strictest(
        self,
        verdicts: list[Verdict],
        all_checks: list,
        avg_score: float,
    ) -> Verdict:
        failed = [v for v in verdicts if not v.passed]
        if not failed:
            return Verdict(
                passed=True,
                reasoning="All reviewers passed",
                severity=SEVERITY_INFO,
                checks=all_checks,
                score=avg_score,
            )
        worst = max(failed, key=lambda v: v.severity.weight)
        reasons = [v.reasoning for v in failed]
        return Verdict(
            passed=False,
            reasoning=f"{len(failed)} reviewer(s) failed: {'; '.join(reasons)}",
            severity=worst.severity,
            checks=all_checks,
            score=avg_score,
        )

    def _resolve_majority(
        self,
        verdicts: list[Verdict],
        all_checks: list,
        avg_score: float,
    ) -> Verdict:
        pass_count = sum(1 for v in verdicts if v.passed)
        passed = pass_count > len(verdicts) / 2
        if passed:
            return Verdict(
                passed=True,
                reasoning=f"Majority passed ({pass_count}/{len(verdicts)})",
                severity=SEVERITY_INFO,
                checks=all_checks,
                score=avg_score,
            )
        failed = [v for v in verdicts if not v.passed]
        worst = max(failed, key=lambda v: v.severity.weight)
        return Verdict(
            passed=False,
            reasoning=f"Majority failed ({len(failed)}/{len(verdicts)})",
            severity=worst.severity,
            checks=all_checks,
            score=avg_score,
        )

    def _resolve_weighted(
        self,
        verdicts: list[Verdict],
        all_checks: list,
    ) -> Verdict:
        total_weight = sum(v.severity.weight for v in verdicts)
        if total_weight == 0:
            return Verdict(
                passed=True,
                reasoning="No weighted signal",
                severity=SEVERITY_INFO,
                checks=all_checks,
                score=1.0,
            )
        weighted_score = sum(v.score * v.severity.weight for v in verdicts) / total_weight
        passed = weighted_score >= 0.5
        worst_failing = [v for v in verdicts if not v.passed]
        severity = (
            max(worst_failing, key=lambda v: v.severity.weight).severity
            if worst_failing
            else SEVERITY_INFO
        )
        return Verdict(
            passed=passed,
            reasoning=f"Weighted score: {weighted_score:.2f}",
            severity=severity,
            checks=all_checks,
            score=weighted_score,
        )
